File: src/python/core/coupling.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
core/coupling.py — 時系列直交結合マトリクス: FFT 相互相関 + 遠ラグ帰無 (Target Echo / E2)
==================================================================================
docs/SPEC_ECHO_GENESIS.md §3.3 / §5.3 / §5.10.3 (W-7〜W-14 最終警告)。

数理の核心 (§2 Note 3): 教科書的な t 検定/p 値は自己相関のある系列に対して嘘を
つく (実効標本数の過大評価)。FFT で全ラグの相互相関を一括計算すると、
信号 (|τ| <= MAX_LAG) と帰無分布 (遠ラグ |τ'| ∈ NULL_LAG_RANGE) が
「同じ 1 回の計算から」手に入る。乱数も外部ライブラリも不要な完全決定論の
有意性判定である。

【W-7〜W-14 (SPEC Rev.3 §5.10.3) を全項目遵守】
  W-7  ゼロパディング不足禁止 (M = 2^ceil(log2(2N)) 未満への「効率化」禁止)
  W-8  n, S_xy, S_x, S_y, S_xx, S_yy の 6 配列レシピは 1 本も省略不可
  W-9  n(τ) は np.rint 後に N_MIN 比較 (irfft は 99.99999997 を返す)
  W-10 分散項は max(·, 0.0) でクランプし、両方 > 1e-9 を有意判定の前提に
  W-11 ρ_ij(τ) = corr(y_i(t), y_j(t+τ))。ラグ符号規約は恒等式でテスト固定
  W-12 帰無ラグにも同じ N_MIN ゲート。有効 null ラグ < 50 は unratable
  W-13 デトレンド/季節調整の追加禁止 (遠ラグ帰無が自動吸収する)
  W-14 タイ破りに乱数 jitter を使わない (平均ランクのみ。I-17 違反防止)
"""

from __future__ import annotations

import numpy as np

MAX_LAG = 14
RHO_MIN = 0.15
N_MIN = 100
NULL_LAG_RANGE = (45, 365)
MIN_NULL_SAMPLES = 50          # W-12: これ未満は帰無分布が信頼できず unratable
MAX_PAIRS_REPORTED = 64


class CouplingError(RuntimeError):
    """coupling_matrix() の入力前提違反 (系列長不足など)。"""


# ---------------------------------------------------------------- §3.2 ランク変換
def _rank_transform(x: np.ndarray, mask: np.ndarray) -> np.ndarray:
    """有効値のみを対象に平均ランク (タイは平均) を付与し、[-1,1] へ線形写像する。
    欠測は 0 (I-18)。単発の外れ値 (突発的な大支出等) が Pearson を支配するのを
    防ぐための Spearman 化。argsort(kind="stable") のみを用い、乱数によるタイ
    破り (W-14) は行わない — 決定論性 (I-17) の要求。"""
    out = np.zeros(x.shape[0], dtype=np.float64)
    valid_idx = np.flatnonzero(mask)
    n = valid_idx.size
    if n < 2:
        return out   # 有効値 1 点以下ではランクが定義できない (分散 0 として扱われる)

    vals = x[valid_idx]
    order = np.argsort(vals, kind="stable")
    ranks = np.empty(n, dtype=np.float64)
    ranks[order] = np.arange(n, dtype=np.float64)

    sorted_vals = vals[order]
    i = 0
    while i < n:
        j = i
        while j + 1 < n and sorted_vals[j + 1] == sorted_vals[i]:
            j += 1
        if j > i:
            avg = ranks[order[i:j + 1]].mean()
            ranks[order[i:j + 1]] = avg
        i = j + 1

    mapped = (ranks / (n - 1)) * 2.0 - 1.0
    out[valid_idx] = mapped
    return out


# ---------------------------------------------------------------- FFT 相互相関
def _circular_xcorr(A: np.ndarray, B: np.ndarray, m: int) -> np.ndarray:
    """(a*b)[tau] = sum_t a[t]*b[t+tau] (circular)。
    = irfft(conj(rfft(a)) * rfft(b))。A,B は事前計算済みの rfft 結果。
    tau は [0, M) の実数配列として返る。負ラグは呼び出し側が負インデックスで
    参照する (numpy の負インデックスがそのまま M+tau を指すため一致する —
    W-11 のラグ符号規約はこの一点に懸かっている)。"""
    return np.fft.irfft(np.conj(A) * B, n=m)


def coupling_matrix(values: np.ndarray, mask: np.ndarray,
                    max_lag: int = MAX_LAG) -> dict:
    """時系列直交結合マトリクス C[i,j,tau] を計算する (SPEC §3.3)。

    values: (N, K) f32/f64。mask: (N, K) bool。K レーンの全ペア (i<j のみ —
    対称性 rho_ij(tau)==rho_ji(-tau) により半分で足りる。W-11) について、
    信号域 |tau|<=max_lag の中から採択条件を満たす最良のラグを報告する。
    採択されない (n_eff 不足で全ラグ unratable な) ペアは省略する。
    """
    values = np.asarray(values, dtype=np.float64)
    mask = np.asarray(mask, dtype=bool)
    if values.shape != mask.shape or values.ndim != 2:
        raise CouplingError(f"values/mask の shape 不一致または非2次元: "
                           f"{values.shape} vs {mask.shape}")
    n_rows, k_lanes = values.shape
    null_lo, null_hi = NULL_LAG_RANGE
    # W-7: ゼロパディングが遠ラグ帰無レンジより短いと、遠ラグ自体が信号域との
    # circular wrap-around を起こし「去年の年末と今年の年始の偽結合」を生む。
    if n_rows <= 2 * null_hi:
        raise CouplingError(
            f"系列長 n_rows={n_rows} は遠ラグ帰無 (|tau|<={null_hi}) を安全に "
            f"取るには短すぎる (最低 {2 * null_hi + 1} 行必要)")

    m = 1 << int(np.ceil(np.log2(2 * n_rows)))   # W-7: 2N 未満への切り詰め禁止

    # 各レーンにつき 3 本の rfft を前計算 (§3.3)
    Y = np.zeros((k_lanes, m // 2 + 1), dtype=np.complex128)
    Mh = np.zeros((k_lanes, m // 2 + 1), dtype=np.complex128)
    Q = np.zeros((k_lanes, m // 2 + 1), dtype=np.complex128)
    for k in range(k_lanes):
        y = _rank_transform(values[:, k], mask[:, k])
        mm = mask[:, k].astype(np.float64)
        Y[k] = np.fft.rfft(y, n=m)
        Mh[k] = np.fft.rfft(mm, n=m)
        Q[k] = np.fft.rfft(y * y, n=m)

    all_lags = np.arange(-max_lag, max_lag + 1)
    null_lags = np.concatenate([
        np.arange(-null_hi, -null_lo + 1),
        np.arange(null_lo, null_hi + 1),
    ])

    def rho_at(S_xy, S_x, S_y, S_xx, S_yy, n_arr, tau: int):
        # W-9: irfft の n(tau) は数学的には整数だが浮動小数の丸め誤差
        # (99.99999997 等) を持つ。N_MIN 比較の前に必ず np.rint する。
        n_round = float(np.rint(n_arr[tau]))
        if n_round < N_MIN:
            return None
        # W-8: この 4 項 (S_x, S_y, S_xx, S_yy) は「グローバル中心化したから
        # 不要」という簡略化を絶対に許さない。省略すると共有欠測パターンが
        # そのまま正の相関として析出する (E1.1 と同じ機序のアーティファクト)。
        var_x = max(n_round * S_xx[tau] - S_x[tau] ** 2, 0.0)   # W-10: 微小負のクランプ
        var_y = max(n_round * S_yy[tau] - S_y[tau] ** 2, 0.0)
        if var_x <= 1e-9 or var_y <= 1e-9:
            # W-10: 0 除算の回避策ではなく「この pair はどちらかの分散が
            # 実質ゼロで評価不能」という情報量の判定そのもの。
            return None
        num = n_round * S_xy[tau] - S_x[tau] * S_y[tau]
        rho = float(np.clip(num / np.sqrt(var_x * var_y), -1.0, 1.0))
        return rho, int(n_round)

    pairs: list[dict] = []
    for i in range(k_lanes):
        for j in range(i + 1, k_lanes):   # 対称性により i<j のみ (W-11)
            S_xy = _circular_xcorr(Y[i], Y[j], m)
            S_x = _circular_xcorr(Y[i], Mh[j], m)
            S_y = _circular_xcorr(Mh[i], Y[j], m)
            S_xx = _circular_xcorr(Q[i], Mh[j], m)
            S_yy = _circular_xcorr(Mh[i], Q[j], m)
            n_arr = _circular_xcorr(Mh[i], Mh[j], m)

            candidates = []
            for tau in all_lags:
                res = rho_at(S_xy, S_x, S_y, S_xx, S_yy, n_arr, int(tau))
                if res is not None:
                    candidates.append((int(tau), res[0], res[1]))
            if not candidates:
                continue   # 全ラグで n_eff 不足 — このペアは記録しない

            # W-12: 帰無ラグにも同一ゲート (N_MIN) を適用する。マスクにより
            # 遠ラグほど n が痩せるため、小標本の高分散 |rho| を帰無に混ぜると
            # q99 が膨張し検出力が死ぬ。
            null_abs = []
            for tau in null_lags:
                res = rho_at(S_xy, S_x, S_y, S_xx, S_yy, n_arr, int(tau))
                if res is not None:
                    null_abs.append(abs(res[0]))
            null_insufficient = len(null_abs) < MIN_NULL_SAMPLES
            null_q99 = (float(np.quantile(null_abs, 0.99, method="linear"))
                       if null_abs else None)

            # 信号域の中から |rho| 最大の候補を選ぶ。all_lags は固定順序の
            # 配列なのでタイの場合も max() は決定論的に先頭要素を返す (I-17)。
            best_tau, best_rho, best_n = max(candidates, key=lambda c: abs(c[1]))

            sig = bool(
                not null_insufficient and null_q99 is not None
                and abs(best_rho) > null_q99
                and abs(best_rho) >= RHO_MIN
                and best_n >= N_MIN
            )
            pairs.append({
                "src": i, "dst": j, "lag": best_tau,
                "rho": round(best_rho, 6), "n_eff": best_n,
                "null_q99": round(null_q99, 6) if null_q99 is not None else None,
                "sig": sig,
                "reason": "null_insufficient" if null_insufficient else None,
            })

    pairs.sort(key=lambda p: -abs(p["rho"]))
    return {"pairs": pairs[:MAX_PAIRS_REPORTED], "n_rows": n_rows,
           "max_lag": max_lag, "n_lanes": k_lanes}

File: apps/desktop/src-tauri/src/analytics/coupling.rs
//! Orthogonal coupling matrix — FFT cross-correlation + far-lag null (Echo E2).
//!
//! Ports `core/coupling.py` contracts (SPEC §3.3 / W-7‥W-14). Pure Rust,
//! no RNG. Missing values use mask only (I-18).

use serde::{Deserialize, Serialize};

pub const MAX_LAG: i32 = 14;
pub const RHO_MIN: f64 = 0.15;
pub const N_MIN: f64 = 100.0;
pub const NULL_LAG_LO: i32 = 45;
pub const NULL_LAG_HI: i32 = 365;
pub const MIN_NULL_SAMPLES: usize = 50;
pub const MAX_PAIRS_REPORTED: usize = 64;

pub const FEATURE_LANES: &[&str] = &[
    "diary_chars",
    "abstract_idx",
    "guilt_idx",
    "productivity_idx",
    "consult_count",
    "spend_total",
    "spend_hedonic",
    "spend_invest",
    "cal_event_count",
    "cal_private_hours",
    "cal_switch_count",
    "line_out_msgs",
    "line_in_msgs",
    "line_out_chars",
    "line_reply_med_min",
    "line_initiations",
    "line_night_out",
    "friction_events",
    "task_declared",
    "task_executed",
    "github_commits",
    "leetcode_solved",
];

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub struct CouplingPair {
    pub src: usize,
    pub dst: usize,
    pub lag: i32,
    pub rho: f64,
    pub n_eff: i32,
    pub null_q99: Option<f64>,
    pub sig: bool,
    pub reason: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub struct CouplingMatrix {
    pub pairs: Vec<CouplingPair>,
    pub n_rows: usize,
    pub max_lag: i32,
    pub n_lanes: usize,
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub enum CouplingError {
    ShapeMismatch,
    SeriesTooShort { n_rows: usize, minimum: usize },
}

impl std::fmt::Display for CouplingError {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            Self::ShapeMismatch => write!(f, "values/mask shape mismatch"),
            Self::SeriesTooShort { n_rows, minimum } => {
                write!(f, "series too short: n_rows={n_rows} need>={minimum}")
            }
        }
    }
}

#[derive(Clone, Copy)]
struct C64 {
    re: f64,
    im: f64,
}

impl C64 {
    fn new(re: f64, im: f64) -> Self {
        Self { re, im }
    }
    fn zero() -> Self {
        Self { re: 0.0, im: 0.0 }
    }
    fn mul(self, o: Self) -> Self {
        Self {
            re: self.re * o.re - self.im * o.im,
            im: self.re * o.im + self.im * o.re,
        }
    }
    fn add(self, o: Self) -> Self {
        Self {
            re: self.re + o.re,
            im: self.im + o.im,
        }
    }
    fn sub(self, o: Self) -> Self {
        Self {
            re: self.re - o.re,
            im: self.im - o.im,
        }
    }
    fn scale(self, s: f64) -> Self {
        Self {
            re: self.re * s,
            im: self.im * s,
        }
    }
    fn conj(self) -> Self {
        Self {
            re: self.re,
            im: -self.im,
        }
    }
}

fn next_pow2(n: usize) -> usize {
    let mut m = 1usize;
    while m < n {
        m <<= 1;
    }
    m
}

/// In-place radix-2 Cooley–Tukey FFT. `invert=true` → IFFT (unnormalized).
fn fft_inplace(a: &mut [C64], invert: bool) {
    let n = a.len();
    debug_assert!(n.is_power_of_two());
    let mut j = 0usize;
    for i in 1..n {
        let mut bit = n >> 1;
        while j & bit != 0 {
            j ^= bit;
            bit >>= 1;
        }
        j ^= bit;
        if i < j {
            a.swap(i, j);
        }
    }
    let mut len = 2;
    while len <= n {
        let ang = std::f64::consts::TAU / (len as f64) * if invert { 1.0 } else { -1.0 };
        let wlen = C64::new(ang.cos(), ang.sin());
        let mut i = 0;
        while i < n {
            let mut w = C64::new(1.0, 0.0);
            for j in 0..(len / 2) {
                let u = a[i + j];
                let v = a[i + j + len / 2].mul(w);
                a[i + j] = u.add(v);
                a[i + j + len / 2] = u.sub(v);
                w = w.mul(wlen);
            }
            i += len;
        }
        len <<= 1;
    }
    if invert {
        let inv = 1.0 / n as f64;
        for x in a.iter_mut() {
            *x = x.scale(inv);
        }
    }
}

fn rfft_pad(signal: &[f64], m: usize) -> Vec<C64> {
    let mut buf = vec![C64::zero(); m];
    for (i, &v) in signal.iter().enumerate() {
        buf[i] = C64::new(v, 0.0);
    }
    fft_inplace(&mut buf, false);
    buf
}

fn circular_xcorr(a_fft: &[C64], b_fft: &[C64], m: usize) -> Vec<f64> {
    let mut prod = vec![C64::zero(); m];
    for i in 0..m {
        prod[i] = a_fft[i].conj().mul(b_fft[i]);
    }
    fft_inplace(&mut prod, true);
    prod.into_iter().map(|c| c.re).collect()
}

/// Average-rank Spearman transform → [-1,1]. Missing → 0 (I-18). No jitter (W-14).
pub fn rank_transform(x: &[f64], mask: &[bool]) -> Vec<f64> {
    let n = x.len();
    let mut out = vec![0.0; n];
    let mut valid_idx: Vec<usize> = (0..n).filter(|&i| mask[i]).collect();
    let nv = valid_idx.len();
    if nv < 2 {
        return out;
    }
    valid_idx.sort_by(|&i, &j| {
        x[i].partial_cmp(&x[j])
            .unwrap_or(std::cmp::Ordering::Equal)
            .then_with(|| i.cmp(&j))
    });
    let mut ranks = vec![0.0_f64; nv];
    for (r, &idx) in valid_idx.iter().enumerate() {
        // placeholder; rewrite with average ties
        let _ = idx;
        ranks[r] = r as f64;
    }
    // Stable argsort ranks with average ties on equal values
    let mut i = 0;
    while i < nv {
        let mut j = i;
        while j + 1 < nv && x[valid_idx[j + 1]] == x[valid_idx[i]] {
            j += 1;
        }
        let avg = (i + j) as f64 / 2.0;
        for k in i..=j {
            ranks[k] = avg;
        }
        i = j + 1;
    }
    let denom = (nv - 1) as f64;
    for (k, &idx) in valid_idx.iter().enumerate() {
        out[idx] = (ranks[k] / denom) * 2.0 - 1.0;
    }
    out
}

fn rho_at(
    s_xy: &[f64],
    s_x: &[f64],
    s_y: &[f64],
    s_xx: &[f64],
    s_yy: &[f64],
    n_arr: &[f64],
    tau: i32,
    m: usize,
) -> Option<(f64, i32)> {
    let idx = if tau >= 0 {
        tau as usize
    } else {
        (m as i32 + tau) as usize
    };
    let n_round = n_arr[idx].round();
    if n_round < N_MIN {
        return None;
    }
    let var_x = (n_round * s_xx[idx] - s_x[idx] * s_x[idx]).max(0.0);
    let var_y = (n_round * s_yy[idx] - s_y[idx] * s_y[idx]).max(0.0);
    if var_x <= 1e-9 || var_y <= 1e-9 {
        return None;
    }
    let num = n_round * s_xy[idx] - s_x[idx] * s_y[idx];
    let rho = (num / (var_x * var_y).sqrt()).clamp(-1.0, 1.0);
    Some((rho, n_round as i32))
}

fn quantile_abs(sorted_abs: &[f64], q: f64) -> f64 {
    if sorted_abs.is_empty() {
        return 0.0;
    }
    let pos = q * (sorted_abs.len() as f64 - 1.0);
    let lo = pos.floor() as usize;
    let hi = pos.ceil() as usize;
    if lo == hi {
        sorted_abs[lo]
    } else {
        let w = pos - lo as f64;
        sorted_abs[lo] * (1.0 - w) + sorted_abs[hi] * w
    }
}

/// `values` / `mask`: row-major (n_rows * n_lanes).
pub fn coupling_matrix(
    values: &[f64],
    mask: &[bool],
    n_rows: usize,
    n_lanes: usize,
    max_lag: i32,
) -> Result<CouplingMatrix, CouplingError> {
    if values.len() != n_rows * n_lanes || mask.len() != values.len() {
        return Err(CouplingError::ShapeMismatch);
    }
    let minimum = (2 * NULL_LAG_HI + 1) as usize;
    if n_rows <= 2 * NULL_LAG_HI as usize {
        return Err(CouplingError::SeriesTooShort {
            n_rows,
            minimum,
        });
    }
    let m = next_pow2(2 * n_rows); // W-7
    let mut y_fft = Vec::with_capacity(n_lanes);
    let mut mh_fft = Vec::with_capacity(n_lanes);
    let mut q_fft = Vec::with_capacity(n_lanes);
    for k in 0..n_lanes {
        let mut col = Vec::with_capacity(n_rows);
        let mut col_mask = Vec::with_capacity(n_rows);
        for t in 0..n_rows {
            col.push(values[t * n_lanes + k]);
            col_mask.push(mask[t * n_lanes + k]);
        }
        let y = rank_transform(&col, &col_mask);
        let mm: Vec<f64> = col_mask.iter().map(|b| if *b { 1.0 } else { 0.0 }).collect();
        let qq: Vec<f64> = y.iter().map(|a| a * a).collect();
        y_fft.push(rfft_pad(&y, m));
        mh_fft.push(rfft_pad(&mm, m));
        q_fft.push(rfft_pad(&qq, m));
    }

    let mut pairs = Vec::new();
    for i in 0..n_lanes {
        for j in (i + 1)..n_lanes {
            let s_xy = circular_xcorr(&y_fft[i], &y_fft[j], m);
            let s_x = circular_xcorr(&y_fft[i], &mh_fft[j], m);
            let s_y = circular_xcorr(&mh_fft[i], &y_fft[j], m);
            let s_xx = circular_xcorr(&q_fft[i], &mh_fft[j], m);
            let s_yy = circular_xcorr(&mh_fft[i], &q_fft[j], m);
            let n_arr = circular_xcorr(&mh_fft[i], &mh_fft[j], m);

            let mut null_abs = Vec::new();
            for tau in -NULL_LAG_HI..=-NULL_LAG_LO {
                if let Some((rho, _)) = rho_at(&s_xy, &s_x, &s_y, &s_xx, &s_yy, &n_arr, tau, m) {
                    null_abs.push(rho.abs());
                }
            }
            for tau in NULL_LAG_LO..=NULL_LAG_HI {
                if let Some((rho, _)) = rho_at(&s_xy, &s_x, &s_y, &s_xx, &s_yy, &n_arr, tau, m) {
                    null_abs.push(rho.abs());
                }
            }
            let (null_q99, null_ok) = if null_abs.len() < MIN_NULL_SAMPLES {
                (None, false)
            } else {
                null_abs.sort_by(|a, b| a.partial_cmp(b).unwrap_or(std::cmp::Ordering::Equal));
                (Some(quantile_abs(&null_abs, 0.99)), true)
            };

            let mut best_full: Option<(f64, i32, i32)> = None;
            for tau in -max_lag..=max_lag {
                if let Some((rho, n_eff)) = rho_at(&s_xy, &s_x, &s_y, &s_xx, &s_yy, &n_arr, tau, m)
                {
                    let better = match best_full {
                        None => true,
                        Some((best_rho, best_lag, _)) => {
                            let a = rho.abs();
                            let b = best_rho.abs();
                            a > b
                                || (a == b && tau.abs() < best_lag.abs())
                                || (a == b && tau.abs() == best_lag.abs() && tau < best_lag)
                        }
                    };
                    if better {
                        best_full = Some((rho, tau, n_eff));
                    }
                }
            }

            if let Some((rho, lag, n_eff)) = best_full {
                let (sig, reason) = if !null_ok {
                    (false, Some("null_insufficient".into()))
                } else {
                    let q = null_q99.unwrap_or(0.0);
                    let sig = rho.abs() > q && rho.abs() >= RHO_MIN;
                    (sig, None)
                };
                pairs.push(CouplingPair {
                    src: i,
                    dst: j,
                    lag,
                    rho: (rho * 1_000_000.0).round() / 1_000_000.0,
                    n_eff,
                    null_q99,
                    sig,
                    reason,
                });
            }
        }
    }

    pairs.sort_by(|a, b| {
        b.rho
            .abs()
            .partial_cmp(&a.rho.abs())
            .unwrap_or(std::cmp::Ordering::Equal)
            .then_with(|| a.src.cmp(&b.src))
            .then_with(|| a.dst.cmp(&b.dst))
    });
    pairs.truncate(MAX_PAIRS_REPORTED);

    Ok(CouplingMatrix {
        pairs,
        n_rows,
        max_lag,
        n_lanes,
    })
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn rank_transform_ties_average() {
        let x = [1.0, 2.0, 2.0, 4.0];
        let m = [true, true, true, true];
        let r = rank_transform(&x, &m);
        assert!((r[1] - r[2]).abs() < 1e-12);
    }
}
